C# module compilation fail with `deprecated=no` compile flag

### Godot version

Master branch, 5ee983188de97ae027f9b9c1443438063f708a7e

### System information

Gentoo - Linux 6.1.57

### Issue description

Following this instructions, https://docs.godotengine.org/en/stable/contributing/development/compiling/compiling_with_dotnet.html#example-linux-bsd with the exception that I was testing with deprecated=no

The compilation fails with these errors:
```
Running MSBuild:  /usr/bin/dotnet msbuild /home/leandrobg/Projects/godot/modules/mono/glue/GodotSharp/GodotSharp.sln /restore /t:Build /p:Configuration=Debug /p:NoWarn=1591
MSBuild version 17.7.3+8ec440e68 for .NET
  Determining projects to restore...
  All projects are up-to-date for restore.
  Godot.SourceGenerators.Internal -> /home/leandrobg/Projects/godot/modules/mono/glue/GodotSharp/Godot.SourceGenerators.Internal/bin/Debug/netstandard2.0/Godot.SourceGenerators.Internal.dll
/home/leandrobg/Projects/godot/modules/mono/glue/GodotSharp/GodotSharp/Compat.cs(33,12): error CS0246: The type or namespace name 'AnimationMethodCallMode' could not be found (are you missing a using directive or an assembly reference?) [/home/leandrobg/Projects/godot/modules/mono/glue/GodotSharp/GodotSharp/GodotSharp.csproj]
/home/leandrobg/Projects/godot/modules/mono/glue/GodotSharp/GodotSharp/Compat.cs(49,12): error CS0246: The type or namespace name 'AnimationProcessCallback' could not be found (are you missing a using directive or an assembly reference?) [/home/leandrobg/Projects/godot/modules/mono/glue/GodotSharp/GodotSharp/GodotSharp.csproj]
/home/leandrobg/Projects/godot/modules/mono/glue/GodotSharp/GodotSharp/Generated/GodotObjects/TextEdit.cs(746,34): warning CS0108: 'TextEdit.TextDirection' hides inherited member 'Control.TextDirection'. Use the new keyword if hiding was intended. [/home/leandrobg/Projects/godot/modules/mono/glue/GodotSharp/GodotSharp/GodotSharp.csproj]
/home/leandrobg/Projects/godot/modules/mono/glue/GodotSharp/GodotSharp/Compat.cs(60,12): error CS0246: The type or namespace name 'AnimationProcessCallback' could not be found (are you missing a using directive or an assembly reference?) [/home/leandrobg/Projects/godot/modules/mono/glue/GodotSharp/GodotSharp/GodotSharp.csproj]
/home/leandrobg/Projects/godot/modules/mono/glue/GodotSharp/GodotSharp/Generated/GodotObjects/RichTextLabel.cs(399,34): warning CS0108: 'RichTextLabel.TextDirection' hides inherited member 'Control.TextDirection'. Use the new keyword if hiding was intended. [/home/leandrobg/Projects/godot/modules/mono/glue/GodotSharp/GodotSharp/GodotSharp.csproj]
/home/leandrobg/Projects/godot/modules/mono/glue/GodotSharp/GodotSharp/Generated/GodotObjects/ArrayMesh.cs(64,32): warning CS0108: 'ArrayMesh.BlendShapeMode' hides inherited member 'Mesh.BlendShapeMode'. Use the new keyword if hiding was intended. [/home/leandrobg/Projects/godot/modules/mono/glue/GodotSharp/GodotSharp/GodotSharp.csproj]
/home/leandrobg/Projects/godot/modules/mono/glue/GodotSharp/GodotSharp/Generated/GodotObjects/GltfAccessor.cs(280,18): warning CS0108: 'GltfAccessor.GetType()' hides inherited member 'object.GetType()'. Use the new keyword if hiding was intended. [/home/leandrobg/Projects/godot/modules/mono/glue/GodotSharp/GodotSharp/GodotSharp.csproj]
/home/leandrobg/Projects/godot/modules/mono/glue/GodotSharp/GodotSharp/Generated/GodotObjects/GltfAccessor.cs(569,43): warning CS0108: 'GltfAccessor.MethodName.GetType' hides inherited member 'object.GetType()'. Use the new keyword if hiding was intended. [/home/leandrobg/Projects/godot/modules/mono/glue/GodotSharp/GodotSharp/GodotSharp.csproj]
/home/leandrobg/Projects/godot/modules/mono/glue/GodotSharp/GodotSharp/Generated/GodotObjects/Button.cs(170,34): warning CS0108: 'Button.TextDirection' hides inherited member 'Control.TextDirection'. Use the new keyword if hiding was intended. [/home/leandrobg/Projects/godot/modules/mono/glue/GodotSharp/GodotSharp/GodotSharp.csproj]
/home/leandrobg/Projects/godot/modules/mono/glue/GodotSharp/GodotSharp/Generated/GodotObjects/Label.cs(244,34): warning CS0108: 'Label.TextDirection' hides inherited member 'Control.TextDirection'. Use the new keyword if hiding was intended. [/home/leandrobg/Projects/godot/modules/mono/glue/GodotSharp/GodotSharp/GodotSharp.csproj]
/home/leandrobg/Projects/godot/modules/mono/glue/GodotSharp/GodotSharp/Generated/GodotObjects/LineEdit.cs(608,34): warning CS0108: 'LineEdit.TextDirection' hides inherited member 'Control.TextDirection'. Use the new keyword if hiding was intended. [/home/leandrobg/Projects/godot/modules/mono/glue/GodotSharp/GodotSharp/GodotSharp.csproj]
/home/leandrobg/Projects/godot/modules/mono/glue/GodotSharp/GodotSharp/Generated/GodotObjects/LinkButton.cs(85,34): warning CS0108: 'LinkButton.TextDirection' hides inherited member 'Control.TextDirection'. Use the new keyword if hiding was intended. [/home/leandrobg/Projects/godot/modules/mono/glue/GodotSharp/GodotSharp/GodotSharp.csproj]
/home/leandrobg/Projects/godot/modules/mono/glue/GodotSharp/GodotSharp/Generated/GodotObjects/MenuBar.cs(77,34): warning CS0108: 'MenuBar.TextDirection' hides inherited member 'Control.TextDirection'. Use the new keyword if hiding was intended. [/home/leandrobg/Projects/godot/modules/mono/glue/GodotSharp/GodotSharp/GodotSharp.csproj]
```

### Steps to reproduce

Compile like this:

1. `scons  p=linuxbsd target=editor deprecated=no  module_mono_enabled=yes`
2. `bin/godot.li nuxbsd.editor.dev.x86_64.llvm.mono --headless --generate-mono-glue modules/mono/glue`
3. `./modules/mono/build_scripts/build_assemblies.py --godot-output-dir=./bin --godot-platform=linuxbsd`

### Minimal reproduction project

N/A

--- a/modules/mono/build_scripts/build_assemblies.py
+++ b/modules/mono/build_scripts/build_assemblies.py
@@ -194,7 +194,7 @@ def run_msbuild(tools: ToolsLocation, sln: str, chdir_to: str, msbuild_args: Opt
     return subprocess.call(args, env=msbuild_env, cwd=chdir_to)
 
 
-def build_godot_api(msbuild_tool, module_dir, output_dir, push_nupkgs_local, precision):
+def build_godot_api(msbuild_tool, module_dir, output_dir, push_nupkgs_local, precision, no_deprecated):
     target_filenames = [
         "GodotSharp.dll",
         "GodotSharp.pdb",
@@ -217,6 +217,8 @@ def build_godot_api(msbuild_tool, module_dir, output_dir, push_nupkgs_local, pre
             args += ["/p:ClearNuGetLocalCache=true", "/p:PushNuGetToLocalSource=" + push_nupkgs_local]
         if precision == "double":
             args += ["/p:GodotFloat64=true"]
+        if no_deprecated:
+            args += ["/p:GodotNoDeprecated=true"]
 
         sln = os.path.join(module_dir, "glue/GodotSharp/GodotSharp.sln")
         exit_code = run_msbuild(msbuild_tool, sln=sln, chdir_to=module_dir, msbuild_args=args)
@@ -336,12 +338,14 @@ def generate_sdk_package_versions():
         f.write(constants)
 
 
-def build_all(msbuild_tool, module_dir, output_dir, godot_platform, dev_debug, push_nupkgs_local, precision):
+def build_all(
+    msbuild_tool, module_dir, output_dir, godot_platform, dev_debug, push_nupkgs_local, precision, no_deprecated
+):
     # Generate SdkPackageVersions.props and VersionDocsUrl constant
     generate_sdk_package_versions()
 
     # Godot API
-    exit_code = build_godot_api(msbuild_tool, module_dir, output_dir, push_nupkgs_local, precision)
+    exit_code = build_godot_api(msbuild_tool, module_dir, output_dir, push_nupkgs_local, precision, no_deprecated)
     if exit_code != 0:
         return exit_code
 
@@ -364,6 +368,8 @@ def build_all(msbuild_tool, module_dir, output_dir, godot_platform, dev_debug, p
         args += ["/p:ClearNuGetLocalCache=true", "/p:PushNuGetToLocalSource=" + push_nupkgs_local]
     if precision == "double":
         args += ["/p:GodotFloat64=true"]
+    if no_deprecated:
+        args += ["/p:GodotNoDeprecated=true"]
     sln = os.path.join(module_dir, "editor/Godot.NET.Sdk/Godot.NET.Sdk.sln")
     exit_code = run_msbuild(msbuild_tool, sln=sln, chdir_to=module_dir, msbuild_args=args)
     if exit_code != 0:
@@ -390,6 +396,12 @@ def main():
     parser.add_argument(
         "--precision", type=str, default="single", choices=["single", "double"], help="Floating-point precision level"
     )
+    parser.add_argument(
+        "--no-deprecated",
+        action="store_true",
+        default=False,
+        help="Build GodotSharp without using deprecated features. This is required, if the engine was built with 'deprecated=no'.",
+    )
 
     args = parser.parse_args()
 
@@ -414,6 +426,7 @@ def main():
         args.dev_debug,
         push_nupkgs_local,
         args.precision,
+        args.no_deprecated,
     )
     sys.exit(exit_code)
 

--- a/modules/mono/glue/GodotSharp/GodotSharp/GodotSharp.csproj
+++ b/modules/mono/glue/GodotSharp/GodotSharp/GodotSharp.csproj
@@ -135,7 +135,7 @@
     <Compile Include="Properties\AssemblyInfo.cs" />
   </ItemGroup>
   <!-- Compat Sources -->
-  <ItemGroup>
+  <ItemGroup Condition=" '$(GodotNoDeprecated)' == '' ">
     <Compile Include="Compat.cs" />
   </ItemGroup>
   <!--

--- a/modules/mono/glue/GodotSharp/GodotSharpEditor/GodotSharpEditor.csproj
+++ b/modules/mono/glue/GodotSharp/GodotSharpEditor/GodotSharpEditor.csproj
@@ -36,7 +36,7 @@
     </ProjectReference>
   </ItemGroup>
   <!-- Compat Sources -->
-  <ItemGroup>
+  <ItemGroup Condition=" '$(GodotNoDeprecated)' == '' ">
     <Compile Include="Compat.cs" />
   </ItemGroup>
   <!--

